Create Security section for Synthetic Tokens (#1234)
UMAprotocol/protocol: documentation/synthetic_tokens/security/security.md

> # Attack Vectors

Reviewer: ```suggestion
# Known Issues
```

Reviewer: I think we should change the name of the file to `Known Issues.md`. 

Also, note the bug in the map.txt! You have this file in `synthetic_tokens/security/security.md` BUT in map.txt you have the file in `synthetic_tokens/security.md`. This will fail. (This is the same bug we had the other day).

I think the file should be `synthetic_tokens/known_issues.md`.

Author: Great, we can make those changes.

> Below is a list of known attack vectors to the proper collateralization of synthetic tokens deployed using this financial contract template. 

Reviewer: ```suggestion
Below is a list of known limitation and attack vectors affecting `ExpiringMultiparty.sol`, the contract used to create synthetic tokens.  
```

> We’ve categorized and ranked the attack vectors by severity. 

Reviewer: Are these ranked by severity? They appear more categorized by type.

Author: If you consider "severity" as whether each issue is more or less impactful to the contract, then attack vectors would appear above open issues, especially those where the conclusion is "assume assumptions not met". 

We can remove this sentence instead.

Note that the financial contract template has an emergency shutdown feature that prematurely expires the contract. 
This can only be triggered by a vote approved by UMA token holders. 
Emergency shutdown prematurely expires the contract and the final resolution depends on a DVM price, which could take 2-4 days to arrive.

> ## Attack vectors affecting incentives to liquidate:

Reviewer: ```suggestion
## Issues affecting incentives to liquidate:
```

### Declining liquidator incentives
* Scenario: In the event of a successful liquidation, the liquidator receives all collateral deposited by the token sponsor, proportional to the size of the position being liquidated. 
As a result, the reward for a liquidator decreases as a token sponsor’s position becomes more undercollateralized. 

> Eventually, if the position is collateralized by < 100%, there is no incentive to liquidate.

Reviewer: ```suggestion
Eventually, if the collateralization ratio for the position is below 100%, there is no incentive to liquidate.
```

> * Solution: Bots should be performant enough that they can capture the  maximum profit from liquidation before the price moves too far. 
> Additionally, UMA will liquidate token sponsors uneconomically early on to support the development of synthetic tokens. 
> In a large market, liquidations should be effected immediately.

Reviewer: ```suggestion
* Solution: Bots should be performant enough that they can capture the  maximum profit from liquidation before the price moves too far; in a large market with attractive arb opportunities we expect competitive forces between bots to trigger liquidations in close-to-real-time. 
In the early days of the system, before large markets develop, UMA will liquidate token sponsors uneconomically to support the development of synthetic tokens. 
```

### Dilution of liquidation rewards to miners
* Scenario: The first person to call liquidate or dispute gets the reward, which results in a gas price war. 
In the limit, the gas price for a liquidation or dispute will be as high as the reward itself, effectively sending the reward to the miner (“miner extractable value”).
* Solution: This is a common problem for all DeFi “arbitrage” transactions. 
One alternative is to use an auction for liquidation to move from a gas bidding war to a price bidding war for the reward.

> # Open Issues
> Below is a list of known open issues regarding the capital efficiency of token sponsor, liquidator, and disputer positions related to this financial contract template. 
> We’ve categorized and ranked the open issues by likelihood of impact. 
> 
> Note that the financial contract template has an emergency shutdown feature that prematurely expires the contract. 
> This can only be triggered by a vote approved by UMA token holders. 
> Emergency shutdown prematurely expires the contract and the final resolution depends on a DVM price, which could take 2-4 days to arrive.
> 
> ## Open issues affecting incentives to liquidate:

Reviewer: ```suggestion
```

### No haircut mechanisms implemented
* Scenario: If any token sponsor’s position is collateralized by < 100%, there is no incentive to liquidate them. 
If the contract expires with any sponsors below 100% collateral, there will be a race to pull collateral from the contract. 
Whichever sponsors and token holders redeem first will receive collateral. 
Whoever comes last will receive less or nothing depending on the amount of collateral left in the contract.
* Solution: Although we expect the economic incentives to work without a haircut mechanism, if a position does reach < 100% collateralization, UMA will liquidate these positions uneconomically.

> ## Open issues affecting capital efficiency: 

Reviewer: ```suggestion
## Issues affecting capital efficiency: 
```

### New token creation
* Scenario: Token sponsors who wish to create new tokens are required to collateralize the new tokens at levels that meet or exceed the global collateralization ratio (GCR). 
However, this makes the role of token sponsor capital inefficient for 2 reasons: 
1. If a token sponsor has an outstanding synthetic token position collateralized well above the GCR, such that the creation of new tokens could be supported by the current amount of excess collateral in their position, they cannot use this collateral to create new synthetic tokens. 
1. If a token sponsor wishes to create new tokens collateralized to a level below the GCR, they must first collateralize their position above the GCR and go through a “slow” withdrawal to withdraw their excess collateral.

> 

Reviewer: What's the solution? We need to say something about how we address this. (The OZ doc does this).

Author: The OZ doc does not seem to have a solution, instead stating: 

Conclusion: Ignore this problem because capital efficiency isn’t a huge concern.

The lack of a solution is why this is listed as an "open issue", rather than an "attack vector" that has a proposed solution. 

Given that we seem to be removing the distinction between these, do you think we should state a "conclusion", as above? I think that would be an appropriate section under each issue that's raised.

Reviewer: I like how you did this.

### Whale griefing attack
* Scenario: A griefing sponsor can add a proportionally large amount of collateral relative to the rest of the contract while creating a very small debt position. 
This drastically imbalances the contract GCR upwards. 
For example, a malicious sponsor can add 100x the current contract collateral while only creating a 1wei position in debt. 
This results in the contract GCR drastically increasing (~100x), making all later sponsors required to be massively over collateralized. 
At best, this slows down future token sponsorship (must repeatedly create and wait on withdraw), and at worst, this can lock all future sponsorship, constricting token supply. 
This could make it difficult or impossible for liquidators to liquidate positions and simultaneously drive up the token price.

> One worrisome attack vector is if a sponsor were to use this to constrict token supply, and then issue a withdraw request to bring their collateralization ratio to 0%. 

Reviewer: this is phrased awkwardly—like it's a second attack vector.

Author: Would you prefer if we rephrased "attack vector" to "edge case"?

If liquidators cannot get tokens to liquidate the sponsor’s position during the liveness period, the sponsor could walk away with a profit. 
There are a number of assumptions that have to be in place for this to work: 
1. Few or no tokens were minted before this sponsor increased the GCR
2. The sponsor is better capitalized than any liquidator
3. A liquidator does not have enough tokens or collateral to issue a partial liquidation large enough to wipe out the profitability of the sponsor. 
If the sponsor is 100x collateralized, a partial liquidation for 1/100ths of the sponsor’s tokens would essentially wipe out all of the profit.

> Any larger liquidations, and they would lose money after the remaining withdrawal went through. 

Reviewer: What's the solution?

Author: As the OZ doc indicates, the solution is to assume that the assumptions listed above are not met. We can clarify this with a "Conclusion" indicating that so a liquid, well-functioning market for synthetic tokens should not meet any of the assumptions. 

From the OZ doc: 

Solution: assume that legitimate token contracts are liquid enough to make this attack impossibly expensive. More analysis needs to be done here around just how liquid a market needs to be to be safe.

> ## Open issues affecting accurate fee payments:

Reviewer: ```suggestion
## Issues affecting accurate fee payments:
```

### Fees may be charged on collateral outside of “Profit From Corruption”
* Scenario: In the current implementation, collateral that isn’t part of PFC is still taxed, which technically isn’t correct. Once a liquidation is settled, any collateral that hasn’t been withdrawn is charged until withdrawn. Similarly, once the contract expires, all non-withdrawn sponsor and tokenholder collateral is charged until withdrawn. 
* Solution: It is up to token holders and liquidators to withdraw collateral they are owed in a timely manner. 

### Final fee for the DVM is determined at liquidation time, not dispute time
* Scenario: The final fee is queried at the time of liquidation, which is not when the price request is issued to the DVM. 
If the final fee were to change between the liquidation and dispute, the contract could technically pay a different final fee than requested by the DVM.
* Solution: The DVM rules should include some flexibility around exactly when fees are queried vs paid. 
The liveness period is generally quite short, so this flexibility doesn’t add much risk.

> ## Open issues affecting ability to withdraw collateral:

Reviewer: ```suggestion
## Issues affecting ability to withdraw collateral:
```

### Network congestion
* Scenario: UMA’s liquidation or dispute bots are down for longer the liveness periods for  liquidations and disputes. 
This causes token sponsors to potentially be improperly liquidated or for the system to become undercollateralized. 

> * Solution: Robust monitoring of liquidator and dispute bots. Increase the liveness period parameter.

Reviewer: These aren't complete sentences, which is out of place with the rest of the doc.

### Rounding errors can accumulate in the contract
* Scenario: All token balance calculations within UMA’s contracts rely on fixed point math using the `FixedPoint` library. 
Due to the limit in decimal precision of this library (and Solidity in general) numerical operations can result in rounding at the least significant digit. 
Rounding errors are not withdrawable by anyone, so they are effectively locked forever.
* Solution: These errors should remain small for now. 

> Long term, the contract could be changed to allow anyone to withdraw the discrepancy at any time, as Uniswap v2 does.

Reviewer: The reference to uniswap v2 feels out of place

Author: Great, we can remove it.
